### PR TITLE
changed close batch checker logic

### LINKED ISSUE #1026: The main_exit_root of L2 is not being updated
Let's assume we have only L1 deposits, to claim it in L2 we need to update the main_exit_root in L2,  but the sequencer only updates it when have L2 txs with batch.
I suggest make an empty batch to update the L2 main_exit_root in the sequencer.

### PR DESCRIPTION
**Pull request by Mikelle**
Closes https://github.com/0xPolygonHermez/zkevm-node/issues/977
Closes https://github.com/0xPolygonHermez/zkevm-node/issues/975
Closes #1026

### What does this PR do?

Changed logic when seq is checking, should it close batch or not

### Reviewers

Main reviewers:

- @arnaubennassar

### CODE REVIEW

File: state/pgstatestorage.go [hunk 1]
```diff
@@ -277,6 +277,55 @@ func (p *PostgresStorage) GetNumberOfBlocksSinceLastGERUpdate(ctx context.Contex
...
+
+	if errors.Is(err, pgx.ErrNoRows) {
+		return ErrNotFound
+	} else if err != nil {
+		return err
+	}
+
+	_, err = e.Exec(ctx, addGlobalExitRootSQL, lastBlockNum, exitRoot.GlobalExitRootNum.String(), exitRoot.MainnetExitRoot, exitRoot.RollupExitRoot, exitRoot.GlobalExitRoot)
+	return err
+}
```
Comment (arnaubennassar) [inline state/pgstatestorage.go:309]: Not sure what's the purpose of this function, who is the consumer?

**Review by arnaubennassar — CHANGES_REQUESTED** [review]

File: state/pgstatestorage.go [hunk 2]
```diff
@@ -749,6 +798,19 @@ func (p *PostgresStorage) closeBatch(ctx context.Context, receipt ProcessingRece
 	return err
 }
 
+// UpdateGERInOpenBatch update ger in open batch
+func (p *PostgresStorage) UpdateGERInOpenBatch(ctx context.Context, ger common.Hash, dbTx pgx.Tx) error {
```
Comment (arnaubennassar) [inline state/pgstatestorage.go:802]: Please add sanity check to ensure that the batch has no txs in it. And force `dbTx != nil`

**Review by arnaubennassar — CHANGES_REQUESTED** [review]

File: state/pgstatestorage.go (hunk 2 above)
Comment (arnaubennassar) [reply state/pgstatestorage.go:802]: Also, it would be nice to not only update the GER, but also the timestamp with `time.Now()`

**Review by arnaubennassar — APPROVED** [review]

File: sequencer/batchbuilder.go [hunk 3]
```diff
@@ -243,23 +243,61 @@ func (s *Sequencer) closeSequence(ctx context.Context) error {
 // in case it's enough blocks since last GER update, long time since last batch and sequence is profitable
 func (s *Sequencer) shouldCloseSequenceInProgress(ctx context.Context) bool {
 	// Check if GER needs to be updated
-	numberOfBlocks, err := s.state.GetNumberOfBlocksSinceLastGERUpdate(ctx, nil)
-	if err != nil && err != state.ErrNotFound {
-		log.Errorf("failed to get last time GER updated, err: %v", err)
+	blockNum, mainnetExitRoot, err := s.state.GetBlockNumAndMainnetExitRootByGER(ctx, s.sequenceInProgress.GlobalExitRoot, nil)
```
Comment (cool-develope) [inline sequencer/batchbuilder.go:246]: it should trigger `object not found` in the init stage.
`GlobalExitRoot = 0x0000...` there is no the related main_exit_root.

File: sequencer/batchbuilder.go [hunk 4]
```diff
@@ -243,23 +243,61 @@ func (s *Sequencer) closeSequence(ctx context.Context) error {
...
-		if len(s.sequenceInProgress.Txs) == 0 {
-			log.Warn("TODO: update GER without closing batch as no txs have been added yet")
+
+	lastGer, err := s.state.GetLatestGlobalExitRoot(ctx, nil)
+	if err != nil {
+		log.Errorf("failed to get latest global exit root, err: %v", err)
+		return false
+	}
+
+	if lastGer.MainnetExitRoot != mainnetExitRoot {
```
Comment (cool-develope) [inline sequencer/batchbuilder.go:258]:
```suggestion
	if lastGer.GlobalExitRoot !=  s.sequenceInProgress.GlobalExitRoot
```
I think the logic should like this...

File: sequencer/batchbuilder.go (hunk 3 above)
Comment (cool-develope) [reply sequencer/batchbuilder.go:246]: `2022-08-11T09:27:08Z    ERROR   sequencer/batchbuilder.go:248   failed to get mainnetExitRoot and blockNum by ger, err: object not found%!(EXTRA string=
/src/log/log.go:106 github.com/0xPolygonHermez/zkevm-node/log.appendStackTraceMaybeArgs()
/src/log/log.go:164 github.com/0xPolygonHermez/zkevm-node/log.Errorf()
/src/sequencer/batchbuilder.go:248 github.com/0xPolygonHermez/zkevm-node/sequencer.(*Sequencer).shouldCloseSequenceInProgress()
/src/sequencer/batchbuilder.go:29 github.com/0xPolygonHermez/zkevm-node/sequencer.(*Sequencer).tryToProcessTx()
/src/sequencer/sequencer.go:132 github.com/0xPolygonHermez/zkevm-node/sequencer.(*Sequencer).Start.func1()
)`

File: state/pgstatestorage.go [hunk 5]
```diff
@@ -749,6 +767,42 @@ func (p *PostgresStorage) closeBatch(ctx context.Context, receipt ProcessingRece
...
+		isBatchHasTxs bool
+	)
+	e := p.getExecQuerier(dbTx)
+	err := e.QueryRow(ctx, getLastBatchNumberSQL).Scan(&batchNumber)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return ErrStateNotSynchronized
+	}
+
+	const isBatchHasTxsQuery = `SELECT raw_txs_data IS NULL FROM state.batch WHERE batch_num = $1`
+	err = e.QueryRow(ctx, isBatchHasTxsQuery).Scan(&isBatchHasTxs)
```
Comment (cool-develope) [inline state/pgstatestorage.go:787]:
```suggestion
	err = e.QueryRow(ctx, isBatchHasTxsQuery, batchNumber).Scan(&isBatchHasTxs)
```

**Review by cool-develope — COMMENTED** [review]
Where is the test part?

File: state/pgstatestorage.go [hunk 6]
```diff
@@ -749,6 +767,42 @@ func (p *PostgresStorage) closeBatch(ctx context.Context, receipt ProcessingRece
...
+	if err != nil {
+		return err
+	}
+
+	if isBatchHasTxs {
```
Comment (cool-develope) [inline state/pgstatestorage.go:792]:
```suggestion
	if !isBatchHasTxs {
```

**Review by cool-develope — APPROVED** [review]

File: state/pgstatestorage.go [hunk 7]
```diff
@@ -749,6 +767,42 @@ func (p *PostgresStorage) closeBatch(ctx context.Context, receipt ProcessingRece
...
+		batchNumber   uint64
+		isBatchHasTxs bool
+	)
+	e := p.getExecQuerier(dbTx)
+	err := e.QueryRow(ctx, getLastBatchNumberSQL).Scan(&batchNumber)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return ErrStateNotSynchronized
+	}
+
+	const isBatchHasTxsQuery = `SELECT EXISTS (SELECT 1 FROM state.batch WHERE raw_txs_data IS NOT NULL AND batch_num = $1)`
```
Comment (arnaubennassar) [inline state/pgstatestorage.go:786]: `raw_txs_data` is only set once the batch is closed. The correct way of checking this is to check if there are l2blocks with the same batch number

**Review by arnaubennassar — CHANGES_REQUESTED** [review]

**Review by arnaubennassar — APPROVED** [review]

**Pull request merged**